zeroex/order.go

```diff
@@ -280,83 +290,128 @@ func (s *SignedOrder) ConvertToOrderWithoutExchangeAddress() wrappers.OrderWitho
+	TakerFee              string `json:"takerFee"`
+	SenderAddress         string `json:"senderAddress"`
+	ExchangeAddress       string `json:"exchangeAddress"`
+	FeeRecipientAddress   string `json:"feeRecipientAddress"`
+	ExpirationTimeSeconds string `json:"expirationTimeSeconds"`
+	Salt                  string `json:"salt"`
+	Signature             string `json:"signature"`
 }
 
 // MarshalJSON implements a custom JSON marshaller for the SignedOrder type
 func (s *SignedOrder) MarshalJSON() ([]byte, error) {
-	return json.Marshal(signedOrderJSON{
-		MakerAddress:          s.MakerAddress,
-		MakerAssetData:        fmt.Sprintf("0x%s", common.Bytes2Hex(s.MakerAssetData)),
+	makerAssetData := ""
+	if len(s.MakerAssetData) != 0 {
+		makerAssetData = fmt.Sprintf("0x%s", common.Bytes2Hex(s.MakerAssetData))
+	}
+	takerAssetData := ""
+	if len(s.TakerAssetData) != 0 {
+		takerAssetData = fmt.Sprintf("0x%s", common.Bytes2Hex(s.TakerAssetData))
+	}
+	signature := ""
+	if len(s.Signature) != 0 {
+		signature = fmt.Sprintf("0x%s", common.Bytes2Hex(s.Signature))
+	}
+
+	signedOrderBytes, err := json.Marshal(SignedOrderJSON{
+		MakerAddress:          s.MakerAddress.Hex(),
+		MakerAssetData:        makerAssetData,
 		MakerAssetAmount:      s.MakerAssetAmount.String(),
 		MakerFee:              s.MakerFee.String(),
-		TakerAddress:          s.TakerAddress,
-		TakerAssetData:        fmt.Sprintf("0x%s", common.Bytes2Hex(s.TakerAssetData)),
+		TakerAddress:          s.TakerAddress.Hex(),
+		TakerAssetData:        takerAssetData,
 		TakerAssetAmount:      s.TakerAssetAmount.String(),
 		TakerFee:              s.TakerFee.String(),
-		SenderAddress:         s.SenderAddress,
-		ExchangeAddress:       s.ExchangeAddress,
-		FeeRecipientAddress:   s.FeeRecipientAddress,
+		SenderAddress:         s.SenderAddress.Hex(),
+		ExchangeAddress:       s.ExchangeAddress.Hex(),
+		FeeRecipientAddress:   s.FeeRecipientAddress.Hex(),
 		ExpirationTimeSeconds: s.ExpirationTimeSeconds.String(),
 		Salt:                  s.Salt.String(),
-		Signature:             fmt.Sprintf("0x%s", common.Bytes2Hex(s.Signature)),
+		Signature:             signature,
 	})
+	return signedOrderBytes, err
 }
 
+const addressHexLength = 42
+
 // UnmarshalJSON implements a custom JSON unmarshaller for the SignedOrder type
 func (s *SignedOrder) UnmarshalJSON(data []byte) error {
-	var signedOrderJSON signedOrderJSON
+	var signedOrderJSON SignedOrderJSON
 	err := json.Unmarshal(data, &signedOrderJSON)
 	if err != nil {
 		return err
 	}
-	s.MakerAddress = signedOrderJSON.MakerAddress
-	s.MakerAssetData = common.Hex2Bytes(signedOrderJSON.MakerAssetData[2:])
+	if len(signedOrderJSON.MakerAddress) >= addressHexLength && signedOrderJSON.MakerAddress[:2] == "0x" {
+		s.MakerAddress = common.HexToAddress(signedOrderJSON.MakerAddress)
+	}
+	if len(signedOrderJSON.MakerAssetData) >= 2 && signedOrderJSON.MakerAssetData[:2] == "0x" {
+		s.MakerAssetData = common.Hex2Bytes(signedOrderJSON.MakerAssetData[2:])
```

Comment:
The [`common.FromHex` function](https://github.com/ethereum/go-ethereum/blob/7fd82a0e3e87aaf26198a6cf2de242c30fa7c2ab/common/bytes.go#L45) already checks for the "0x" prefix if you just want to use that instead.

Reply:
Hm, it's not as strict as I would like. I don't think we should allow clients to send `MakerAssetData` without the `0x` prefix, or with a `0X` prefix. Let's leave it as is for now.

Reply:
> I don't think we should allow clients to send MakerAssetData without the 0x prefix

Shouldn't that be part of the JSON Schema? Or is it not possible to enforce that constraint?

Reply:
I just checked and it looks like it is already enforced by the JSON schema.